### PR TITLE
Update

### PR DESCRIPTION
**Pull request by tomaskulich**

### CODE REVIEW

File: src/index.adoc [hunk 1]
```diff
@@ -4,14 +4,542 @@ include::header.adoc[]
...
+}
+```
+
+Don't get trapped by dogma here: many tasks can not be easily addressed using only 100% pure functions.
+Be ready to conduct some side-effects, read (and sometimes even modify) variables from the outer scopes,
+etc. The learning here is, that you should pursue the purity to the reassonable level.
+
+As an example, Haskell takes purity really seriously and perform side-effects by monads. Monads
+seemingly make all functions pure, but in reality, it's the kind of purity which costs you a lot,
+while not buying you either better testability, nor it ease your resoning about the code.
```
Comment (rtrembecky) [inline src/index.adoc]:
```suggestion
seemingly make all functions pure, but in reality, it's the kind of purity that costs you a lot -
while not buying you better testability, it doesn't make reasoning about the code easier either.
```

File: src/index.adoc [hunk 2]
```diff
@@ -4,14 +4,542 @@ include::header.adoc[]
...
+
+=== Prefer simple data structures
+
+In Java-ish languages one does often define structures like this:
```
Comment (rtrembecky) [inline src/index.adoc]:
```suggestion
In Java-ish languages one often defines structures like this:
```

File: src/index.adoc [hunk 3]
```diff
@@ -4,14 +4,542 @@ include::header.adoc[]
...
+  }
+}
+
+const me = new Person(...)
+```
+
+In JS, TS, Clojure the second approach is very common:
+
+```
+const me:User = {first_name: "Tomas", last_name: "Kulich"}
```
Comment (rtrembecky) [inline src/index.adoc]:
```suggestion
const me: User = {first_name: "Tomas", last_name: "Kulich"}
```

File: src/index.adoc [hunk 4]
```diff
@@ -4,14 +4,542 @@ include::header.adoc[]
...
+there are several easy ways how to update the object in an immutable manner:
+
+```
+const updated_user = {...user, name: new_name}
+// or this one, ideal for nested objects
+const updated_user = immutability_helper.update(user, {name: {$set: new_name}})
+```
+There exist immutable data structures implementation (Map, List) that do this in a very performant
+manner (time complexity `O(lg(structure_size))`), however, unless the structures are huge, you can
+just stick to good-old mutable objects and just not mutating them by convention.
```
Comment (rtrembecky) [inline src/index.adoc]:
```suggestion
just stick to the good-old mutable objects and just avoid mutating them by convention.
```

File: src/index.adoc [hunk 5]
```diff
@@ -4,14 +4,542 @@ include::header.adoc[]
...
+=== Stay pragmatic
+
+There are many wrong things to optimize for:
+- extreme purity,
+- coding the idiomatic [your favorite framework] way,
+- 100% test coverage,
+- your code behaving 100% correctly even in unlike and unimportant situations, etc.
+
+We're solving problems. Focus on "what's the problem we're trying to solve here" rather than "what's
+the most idiomatic way to achive this in framework X".
```
Comment (rtrembecky) [inline src/index.adoc]:
```suggestion
the most idiomatic way to achieve this in framework X".
```

File: src/index.adoc [hunk 6]
```diff
@@ -4,14 +4,542 @@ include::header.adoc[]
...
+
+This doesn't advocate writing shitty code. Saving 30 minutes of your time just to create a day long
+debug session for yourself is certainly not pragmatic.
+
+=== Avoid premature optimization and premature generalizations
+
+Always be aware to how much resources (time, memory) you can waste, so you can waste them visely.
```
Comment (rtrembecky) [inline src/index.adoc]:
```suggestion
Always be aware of how much resources (time, memory) you can waste, so you can waste them wisely.
```

File: src/index.adoc [hunk 7]
```diff
@@ -4,14 +4,542 @@ include::header.adoc[]
...
+With time, be aware of what's the most time-consuming thing that's happening and ignore everything
+else, prefering readability to unimportant time-saving. If a reviewer asks you to optimize O(n) work
```
Comment (rtrembecky) [inline src/index.adoc]:
```suggestion
else, preferring readability to unimportant time-saving. If a reviewer asks you to optimize O(n) work
```

File: src/index.adoc [hunk 8]
```diff
@@ -4,14 +4,542 @@ include::header.adoc[]
...
+differently: they quickly produce 100 lines of bad code, then using print and debugger to get the
+code in shape. Not only this feels worse, in the end, it's also much more time consuming.
+
+If you fall into the second case, consider changing your workflow. Slow down and focus on writing
+high-quality code. Reason why the code should work ("On this place, the connection must be already
+opened, because I jump here only from here and here... wait, also from here! Oh my, that'd be a
+nasty bug to have."). Play a "coding game": If the code runs correctly on the first time (possibly
+after fixing small and easy bugs such as typos), you've won. Soon you develop a healthy ammount of
+paranoia, which'll make you question your code frequently. You'll produce a code 3x slower but also
+10x faster.
```
Comment (rtrembecky) [inline src/index.adoc]:
```suggestion
Be very humble when predicting what new features the business will need. Even if you're right that X
will be needed in 2 months, the preferred way to achieve X is likely to change by that time. And
even if you're completely right about all your predictions, it still may be easier to postpone X for
later, since you won't need to maintain X for 2 extra months.

=== Produce bug-free code by avoiding bugs

Imagine you write 100 lines. You hit run and--it's working! What a feeling! Some people do it
differently: they quickly produce 100 lines of bad code, then using print and debugger to get the
code in shape. Not only this feels worse, in the end, it's also much more time-consuming.

If you fall into the second case, consider changing your workflow. Slow down and focus on writing
high-quality code. The reason why the code should work ("On this place, the connection must be already
opened, because I jump here only from here and here... wait, also from here! Oh my, that'd be a
nasty bug to have."). Play a "coding game": If the code runs correctly on the first try (possibly
after fixing small and easy bugs such as typos), you've won. Soon you develop a healthy amount of
paranoia, which'll make you question your code frequently. You'll produce a code 3x slower but also
10x faster.
```

File: src/index.adoc [hunk 9]
```diff
@@ -4,14 +4,542 @@ include::header.adoc[]
...
+- load an object from DB,
+- assert that all attributes are correctly loaded.
+
+The test has a low value and probably doesn't pull its weight:
+- it's testing a relatively simple code; it'd be easier to review it manually,
+- the code is not likely to break,
+- if test passes, it doesn't mean much; there are many edge cases (e.g. very long user name), that are
+  not covered,
+- if test fails, you'd probably find out anyway really,
+- it doesn't serve as a documentation to something worth documenting.
```
Comment (rtrembecky) [inline src/index.adoc]:
```suggestion
How to write tests is almost a religious thingy, with several thought schools. Stay away from the
dogmatic mess they've created; simple pragmatism is the best guidance. If the test adds value that
outweighs the cost (writing, maintaining, and running the test), then it's a good test to have.

Imagine such a test:
- you populate a User object with data,
- save the object to the database,
- load an object from DB,
- assert that all attributes are correctly loaded.

The test has a low value and probably doesn't pull its weight:
- it's testing a relatively simple code; it'd be easier to review it manually,
- the code is not likely to break,
- if the test passes, it doesn't mean much; there are many edge cases (e.g. very long user name), that are
  not covered,
- if the test fails, you'd probably find out anyway really,
- it doesn't serve as documentation to something worth documenting.
```

File: src/index.adoc [hunk 10]
```diff
@@ -4,14 +4,542 @@ include::header.adoc[]
...
+
+=== Code reviews are awesome
+
+They are the best tool we have to produce a bug-free, readable and maintainable code. Often, more
+than one review is necessary for a code to be merged.
+
+Few tips:
+
+- Don't enforce your taste to others too much. If you know shorter form of how to write something
+  but you're not sure, just give advice: 'consider rewriting to ...`.
```
Comment (rtrembecky) [inline src/index.adoc]: `'` vs `` ` ``

File: src/index.adoc [hunk 11]
```diff
@@ -4,14 +4,542 @@ include::header.adoc[]
...
+- This sucks: correcting 10 OKish namings just because it's easy to do and missing a huge design
+  flaw at the same time.
+- If you've had an objective problem understanding something and author had to explain it to you, it
+  probablyu deserves a comment.
```
Comment (rtrembecky) [inline src/index.adoc]:
```suggestion
  probably deserves a comment.
```

File: src/index.adoc [hunk 12]
```diff
@@ -4,14 +4,542 @@ include::header.adoc[]
...
+- If it's for a longer discussion, schedule a call or just pick up the phone and call the other
+  person.
+
+Finally, if it's not clear why the code works correctly, you shouldn't merge it. Author can use
+different ways to persuade you, she did a good job:
```
Comment (rtrembecky) [inline src/index.adoc]:
```suggestion
different ways to persuade you she did a good job:
```

File: src/index.adoc [hunk 13]
```diff
@@ -4,14 +4,542 @@ include::header.adoc[]
...
+- writing tests, demonstrating the desired functionality,
+- structuring a code in a way, that certain behavior is enforced by design,
+- Q: "hey, I'm not sure about this", A: "Me neither, but I've tested it manually and it seems to
+  work. It's poorly documented, so it'll do for now I guess."
+
+The code review process is not so easy to get right.
+
+- Reviewer comments to the PR.
+- Author reacts :thumbsup:, if he accepts the proposed change. If it's something more complex, the
+  discussion can take place. Avoid writing a lot it's inefficient and frustrating. Have a quick call
```
Comment (rtrembecky) [inline src/index.adoc]:
```suggestion
  discussion can take place. Avoid writing a lot, it's inefficient and frustrating. Have a quick call
```

File: src/index.adoc [hunk 14]
```diff
@@ -4,14 +4,542 @@ include::header.adoc[]
...
+  instead.
+- Reviewer checks that she likes the outcome by clicking "mark as resolved".
+- If reviewer has only cosmettic comments, he can approve the PR indicating he doesn't care that
```
Comment (rtrembecky) [inline src/index.adoc]:
```suggestion
- If reviewer has only cosmetic comments, she can approve the PR indicating he doesn't care that
```

File: src/index.adoc [hunk 15]
```diff
@@ -4,14 +4,542 @@ include::header.adoc[]
...
+- Resolving the reviewer's comment by the author in a case where author is not 100% sure he fixed it
+  in a way reviewer intended.
+
+The changes are pushed either as a new "fixing comments" commit, or the history is editted.
+Adapt the customs of your project.
+
+=== Microservices are often overused
+
+There's a too big hype around microservices. Obviously, the world doesn't run as one monolithic
+service (and that's a good think). Since "big project" ==  "small world", it's not a good idea for a
```
Comment (rtrembecky) [inline src/index.adoc:467]:
```suggestion
service (and that's a good thinking). Since "big project" == "small world", it's not a good idea for a
```

File: src/index.adoc [hunk 16]
```diff
@@ -4,14 +4,542 @@ include::header.adoc[]
...
+big project to run as a monolithic service either. Structuring the project to smaller microservices
+helps--as long as the microservices have well-defined responsibilities and the cross-cutting
+concerns are limitted.
+
+Unhealthy hype about microservices gave birth to systems with too much microservices and a lot of
```
Comment (rtrembecky) [inline src/index.adoc]:
```suggestion
Unhealthy hype about microservices gave birth to systems with too many microservices and a lot of
```

File: src/index.adoc [hunk 17]
```diff
@@ -4,14 +4,542 @@ include::header.adoc[]
...
+have a lot of tooling built around it, and as your system evolves, you'll probably need to build
+some custom tooling as well.
+
+=== Fat code is good
+When you start coding, you name your variables `i`, `j`, `k`. Then you code a project with 10k lines
+and you discover using better names such as: `user`, `account`, `ammount` or `transfer()` is good
+for your sanity.
+Then you code a project with 100k lines, discovering you need even better names:
+`account_sender_or_system_account`,
+`ammount_USD_before_VAT`, `try_transfer_with_SEPA()`. Often people use
```
Comment (rtrembecky) [inline src/index.adoc]:
```suggestion
`amount_USD_before_VAT`, `try_transfer_with_SEPA()`. Often people use
```

File: src/index.adoc [hunk 18]
```diff
@@ -4,14 +4,542 @@ include::header.adoc[]
...
+short and imprecise names that kick ass badly, once the project grows.
+
+Good type system can help you shorten the names of your variables and parameters, since a lot of
+info is just to describe the type. This is not always possible, especially for functions you should
+be quite verbose.
+
+Using long names results in writing a fat code: simple expression such as `a = b + c + d` can overflow the
+100 characters limit, which looks ugly and not that readable. This is usually less severe
+problem as one may think.
```
Comment (rtrembecky) [inline src/index.adoc]:
```suggestion
100 characters limit, which looks ugly and not that readable. However, this is usually a less severe
problem than one may think.
```

File: src/index.adoc [hunk 19]
```diff
@@ -4,14 +4,542 @@ include::header.adoc[]
...
+
+As a rule of thumb, when explaining more complected or suspicious code to the curious reviewer,
+consider adding a comment such as:
+
+```
+// Parameter `user` is actually required. It's marked optional just because of tests.
+```
+
+Another case, that always deserves a comment: copy-pasting code from somewhere. Always add a link to
+where the code came from.
```
Comment (rtrembecky) [inline src/index.adoc:526]: I love these two tips. I can't stress them enough myself.

File: src/index.adoc [hunk 20]
```diff
@@ -4,14 +4,542 @@ include::header.adoc[]
...
+
+=== Idempotency
+
+Imagine a service that allows you to create new tansactions as such:
```
Comment (rtrembecky) [inline src/index.adoc]:
```suggestion
Imagine a service that allows you to create new transactions as such:
```

File: src/index.adoc [hunk 21]
```diff
@@ -4,14 +4,542 @@ include::header.adoc[]
...
+
+```
+external_service.add_new_transaction(from: "john", to: "jane", usd_ammount: 10)
```
Comment (rtrembecky) [inline src/index.adoc]:
```suggestion
external_service.add_new_transaction({from: "john", to: "jane", usd_amount: 10})
```

File: src/index.adoc [hunk 22]
```diff
@@ -4,14 +4,542 @@ include::header.adoc[]
...
+  external_service.add_new_transaction({from_account: "john", to_account: "jane", usd_ammount: 10})
+} catch (err) {
+  // what now?
+}
+```
+
+How would you handle the error? You may want to repeat the transaction:
+
+```
+function try_send_transaction_with_repeat({from_account, to_account, usd_ammount, try_this_many_times}) {
```
Comment (rtrembecky) [inline src/index.adoc]:
```suggestion
function try_send_transaction_with_repeat({from_account, to_account, usd_amount, try_this_many_times}) {
```

File: src/index.adoc [hunk 23]
```diff
@@ -4,14 +4,542 @@ include::header.adoc[]
...
+
+```
+external_service.add_new_transaction(from: "john", to: "jane", usd_ammount: 10)
+```
+
+To handle possible errors, you need to add some try-catch logic:
+
+```
+try {
+  external_service.add_new_transaction({from_account: "john", to_account: "jane", usd_ammount: 10})
```
Comment (rtrembecky) [inline src/index.adoc]:
```suggestion
  external_service.add_new_transaction({from_account: "john", to_account: "jane", usd_amount: 10})
```

File: src/index.adoc [hunk 24]
```diff
@@ -4,14 +4,542 @@ include::header.adoc[]
...
+```
+function try_send_transaction_with_repeat({from_account, to_account, usd_ammount, try_this_many_times}) {
+  // default value
+  try_this_many_times = try_this_many_times || 10
+  try {
+    external_service_do_your_job
+  } catch (err) {
+    if (try_this_many_times > 0) {
+      wait(10000)
+      try_send_transaction({from_account, to_account, usd_ammount, try_this_many_times - 1})
```
Comment (rtrembecky) [inline src/index.adoc]:
```suggestion
      try_send_transaction({from_account, to_account, usd_amount, try_this_many_times - 1})
```

File: src/index.adoc [hunk 25]
```diff
@@ -4,14 +4,542 @@ include::header.adoc[]
...
+If you call this with `try_this_many_times === 10`, you'll get 10 tries in 10s intervals and then
+error is thrown. But what if try_send_transaction actually _created_ the transaction on the external
+service and just then something failed, when sending the response back (e.g. parsing the response on
+your server is faulty)? You'd end up with 10 transaction being sent: uh-oh.
+
+To overcome this, decent external services let you specify so-called _idempotency key_ which serves
+as unique identifiers to your request. If the external service is decent, you can use it as such:
+
+```
+function try_send_transaction_with_repeat({from_account, to_account, usd_ammount, try_this_many_times, idempotency_key}) {
```
Comment (rtrembecky) [inline src/index.adoc]:
```suggestion
function try_send_transaction_with_repeat({from_account, to_account, usd_amount, try_this_many_times, idempotency_key}) {
```

**Review by rtrembecky — APPROVED** [review]
looking good, though I found several typos.
also, to be even more grammatically correct, many articles (`the`, `a/an`) should be added, but whatever 😄 or some questionable commas... it would be nice to run it through some grammar-checking software (I fixed some issues using the Grammarly extension - though it doesn't really help with logical commas much)

File: src/index.adoc [hunk 26]
```diff
@@ -4,14 +4,542 @@ include::header.adoc[]
...
+more powerful than ad-hoc trying random stuff with a strong debugger.
 
 == Tools
-include::tools/index.adoc[]
+
+=== Operating System
+Use either Mac or Linux for development. Don't use Windows, don't even think
+about it! There is always something working differently on Windows than on the
+other platforms. This costs us huge amounts of time debugging problems instead
+of developing valuable features.
```
Comment (szist) [inline src/index.adoc:80]: Well. I'm not sure it counts, quite a few people I know work from WSL = windows subsystem for linux. You basically get a virtualized linux inside of Windows, but the layer is very lightweight and you don't loose too much performance. Big bonus is that you avoid most of the driver issues.

Maybe would change it to "native" Windows, or mention that WSL2 is actually fine. I'm not saying I would recommend it over mac or linux, but it's much better than wasting days on broken laptop drivers and bad unstable UX that's often the case on modern linux desktop environments.

File: src/index.adoc [hunk 27]
```diff
@@ -4,14 +4,542 @@ include::header.adoc[]
...
+options link:https://brew.sh[Homebrew^] and
+link:https://www.macports.org[MacPorts^]. The most used in our company is
+Homebrew.
+
+=== Terminal
+
+Learn link:http://linuxcommand.org/learning_the_shell.php[the basics of UNIX shell^].
+
+Replace your default shell with much better one called _zsh_. Install
+link:https://ohmyz.sh/[oh-my-zsh^], where you can select some nice visual theme for your shell.
```
Comment (szist) [inline src/index.adoc:107]: Imho for beginners, bash should be the standard. Most guides and tools work really well with that. With zsh even on mac, a lot of tools have hiccups, bad installations, bad env setup. On linux it made me switch away from it. Last time we ran into that was with `volta.sh` . You can install powerline for bash as well.
https://github.com/powerline/powerline
https://medium.com/earlybyte/powerline-for-bash-6d3dd004f6fc

File: src/index.adoc [hunk 28]
```diff
@@ -4,14 +4,542 @@ include::header.adoc[]
...
+=== Text Editor
+
+==== Visual Studio Code
+
+image::editors/screenshot-vscode.png[]
+
+link:https://code.visualstudio.com[Visual Studio Code site]
+
+Open source editor by Microsoft. It comes prepared for web-development with
+integrated support for JavaScript intellisense, git integration and with many
```
Comment (szist) [inline src/index.adoc]: +Typescript

File: src/index.adoc [hunk 29]
```diff
@@ -4,14 +4,542 @@ include::header.adoc[]
...
+Learning JS React & Redux::
+We use cool modern JavaScript technology like
+link:http://facebook.github.io/react/[React] and link:http://redux.js.org[Redux].
+
+Learning Python::
+We use link:http://www.diveintopython3.net[Python3] with Django.
+Have a look at this great link:http://gettingstartedwithdjango.com[Django tutorial].
+
+Clojure intro::
+http://www.braveclojure.com/foreword/
```
Comment (szist) [inline src/index.adoc:230]: We should replace this with Kotlin probably.

File: src/index.adoc [hunk 30]
```diff
@@ -4,14 +4,542 @@ include::header.adoc[]
...
+
+=== Operating System
+Use either Mac or Linux for development. Don't use Windows, don't even think
+about it! There is always something working differently on Windows than on the
+other platforms. This costs us huge amounts of time debugging problems instead
+of developing valuable features.
+
+If you are new to Linux, install link:http://www.ubuntu.com[Ubuntu^]. Make sure to read
+link:https://help.ubuntu.com/stable/ubuntu-help[the introduction to Ubuntu Desktop^], especially
+link:https://help.ubuntu.com/stable/ubuntu-help/addremove.html[how the software installation work in Linux^].
```
Comment (szist) [inline src/index.adoc:84]: I would even go as far as recommend manjaro over Ubuntu for newcomers. Better installation, better package manager with more packages out of the box. The default DEs KDE/XFCE >>> gnome for anyone new to linux. (gnome a very unintuitive UX for anyone coming from windows)

**Comment by rtrembecky** [comment]
feel free to resolve my comments on typos if you fixed them, I don't plan on going through it again... however, "amount" is really not spelled "ammount" 😄

File: src/index.adoc (hunk 30 above)
Comment (zatkowich) [reply src/index.adoc:84]: Well, I would recommend against Manjaro. It's a rolling distro and don't forget that can be often tricky even for power users. I am, well, not a Linux guru but a long time user that tried many distros.

Manjaro is maybe for example a good gaming distro but in my year-long experience it got broken a few times and I had to fix it manually.
There is AUR which is good for users like me but stuff there is often broken too and one has to edit install scripts. Moreover the packages there work against Arch packages and Manjaro is 1-2 weeks late compared to Arch - and that breaks stuff too.

Yes, many of us are against Ubuntu because of various reasons but hey, it's still a good, "simple" distro for beginners together with Mint.
They can easily switch to their own choice when they would be capable of doing that.

Comment (kravemir) [reply src/index.adoc:84]: I had been using ArchLinux in past, and there was some amount of time spent due to things being broken by updates. Since I moved to Ubuntu, maintenance costs / time spent was heavily reduced. I am not heavily chasing performance anymore, though.

If I was to switch to something else, which would be more complex than Ubuntu, then I would consider Qubes OS. It would give me more value (increased security) for the extra maintenance cost. It might even reduce costs in long-term, as it allows to separate project specific tools, and easily drop them completely (delete qube / virtual machine) without a need to reinstall OS to be sure there aren't any leftovers.

File: src/index.adoc (hunk 26 above)
Comment (kravemir) [reply src/index.adoc:80]: > ... broken laptop drivers ...

No issues with Intel and AMD gpus. I'm not gaming anymore, though. For office usage it works smoothly.

> ... bad unstable UX that's often the case on modern linux desktop environments. ...

The "bad UX" is quite subjective, it more depends on habits. I'm fluently working with GNOME 3 with few extensions, and keybindings to switch to other workspaces.

As of the "unstable" part, well I haven't seen this since I moved from ArchLinux to Ubuntu and I'm using only LTS releases.

Comment (szist) [reply src/index.adoc:80]: It's not the gpus I worry about too much, that gets enough visibility. I haven't had a notebook that didn't have issues with at least one of these things: wifi, bluetooth, soundcard, sleep, touchpad, thunderbolt, webcam, keyboard shortcuts, touchpad gestures (multitouch), touchscreen or webcam (although lately the list is usually shrinking). Our lenovos I think are doing quite OK though apart from touch :).

File: src/index.adoc [hunk 31]
```diff
@@ -4,14 +4,570 @@ include::header.adoc[]
...
+feature branch should be squashed to a small number of commits (often just one) before merging.
+
+Another approach is: you can put more than one atomic change into a branch. In such a case, one
+atomic change gets represented by one commit. The reviewer reviews the branch commit-by-commit and
+when she requires change, author edits the history. Reviewer doesn't see the correcting diff itself,
+but can then check out her comments, the current state of the code and see, if the code looks good
+now. Pros: it works better for doing several tasks simultaneously. For example, you start coding a
+feature, refactor something on-the-fly and fix a bug you discover. It's much easier to have just one
+branch for that. Cons: reviewer doesn't see an exact diff, that author produced to implement their
+change.
```
Comment (gyorgy-kurucz-vacuum) [inline src/index.adoc:459]: > author edits the history. Reviewer doesn't see the correcting diff

A solution actually exists, `git range-diff`, though not many people seem to use it. I often find it useful in exactly this kind of situation: verifying some small change to a long commit series.
